transmission_4: init at 4.0.3

ajs124 opened the pull request:
###### Description of changes
https://github.com/transmission/transmission/releases/tag/4.0.0

###### Things done
- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

ajs124 committed:
transmission: 3.0.0 → 4.0.2

https://github.com/transmission/transmission/releases/tag/4.0.0
https://github.com/transmission/transmission/releases/tag/4.0.1
https://github.com/transmission/transmission/releases/tag/4.0.2

Not adding fmt and utf8cpp to AppArmor list since those are header only.

Co-Authored-By: Jan Tojnar <jtojnar@gmail.com>

ajs124 commented:
seems to break `fragments` and `torrential`

ajs124 commented:
`fragments` (even after updating to 2.1) seems incompatible with this release of transmission:
```
 ERROR transmission_client::client > Unable to parse json: invalid type: boolean `true`, expected i32 at line 1 column 6015
```
cc @emilytrau

mlyxshi commented:
According to [Building-Transmission](https://github.com/transmission/transmission/blob/main/docs/Building-Transmission.md), I think we can use `fetchSubmodules` to fetch dependency. 

I only use `transmission-daemon`, `transmission-remote`, and `web html`, https://github.com/mlyxshi/flake/blob/main/pkgs/transmission/default.nix

ruuda commented:
Result of `nixpkgs-review` on x86_64-linux:

```
2 packages failed to build:
fragments torrential

5 packages built:
dht libtransmission transmission transmission-gtk transmission-qt
```

winterqt commented:
@mlyxshi

> According to [Building-Transmission](https://github.com/transmission/transmission/blob/main/docs/Building-Transmission.md), I think we can use `fetchSubmodules` to fetch third-party dependencies so that we don't need to keep track of them in Nixpkgs. 

We prefer not using vendored libraries whenever possible, and since things work fine with our versions, I don't see why we should change that all of a sudden.

> I only use `transmission-daemon`, `transmission-remote`, and `web html`, https://github.com/mlyxshi/flake/blob/main/pkgs/transmission/default.nix

Is there something you forgot to say here? (I ask because I don't see its relevance, and want to make sire this bump doesn't regress anything.)

vcunat commented:
- [ ] I think it would be nice to fix `*-darwin` (see ofBorg), as they currently build in nixpkgs ([86](https://hydra.nixos.org/job/nixpkgs/trunk/transmission.x86_64-darwin), [aarch](https://hydra.nixos.org/job/nixpkgs/trunk/transmission.aarch64-darwin)).

winterqt commented:
Sure, I'll look into it some time in the next few hours, looks simple enough.

winterqt commented:
```diff
diff --git a/pkgs/applications/networking/p2p/transmission/default.nix b/pkgs/applications/networking/p2p/transmission/default.nix
index 80e159d9545..07a82bbc24c 100644
--- a/pkgs/applications/networking/p2p/transmission/default.nix
+++ b/pkgs/applications/networking/p2p/transmission/default.nix
@@ -32,6 +32,7 @@
 , enableCli ? true
 , installLib ? false
 , apparmorRulesFromClosure
+, Foundation
 }:
 
 let
@@ -62,6 +63,12 @@ in stdenv.mkDerivation {
       "-DENABLE_DAEMON=${mkFlag enableDaemon}"
       "-DENABLE_CLI=${mkFlag enableCli}"
       "-DINSTALL_LIB=${mkFlag installLib}"
+    ] ++ lib.optionals stdenv.isDarwin [
+      # Transmission sets this to 10.13 if not explicitly specified, see https://github.com/transmission/transmission/blob/0be7091eb12f4eb55f6690f313ef70a66795ee72/CMakeLists.txt#L7-L16.
+      "-DCMAKE_OSX_DEPLOYMENT_TARGET=${stdenv.hostPlatform.darwinMinVersion}"
+
+      # This doesn't work on Darwin, see #214945.
+      "-DRUN_CLANG_TIDY=OFF"
     ];
 
   nativeBuildInputs = [
@@ -90,7 +97,7 @@ in stdenv.mkDerivation {
   ++ lib.optionals enableGTK3 [ gtk3 gtkmm3 xorg.libpthreadstubs ]
   ++ lib.optionals enableSystemd [ systemd ]
   ++ lib.optionals stdenv.isLinux [ inotify-tools ]
-  ++ lib.optionals stdenv.isDarwin [ libiconv ];
+  ++ lib.optionals stdenv.isDarwin [ libiconv Foundation ];
 
   postInstall = ''
     mkdir $apparmor
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index 72a973650f7..1c7bef1076b 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -33334,7 +33334,9 @@ with pkgs;
 
   transcribe = callPackage ../applications/audio/transcribe { };
 
-  transmission = callPackage ../applications/networking/p2p/transmission { };
+  transmission = darwin.apple_sdk_11_0.callPackage ../applications/networking/p2p/transmission {
+    inherit (darwin.apple_sdk_11_0.frameworks) Foundation;
+  };
   libtransmission = transmission.override {
     installLib = true;
     enableDaemon = false;
```

ajs124 commented:
What does the `darwin.apple_sdk_11_0` in `all-packages.nix` do and is that really necessary? I'd assume/hope the `inherit` is enough.

The inputs are sorted… somehow? I just added `gtkmm3` there, because that's where the other gtk stuff is.

winterqt commented:
> Edit: apparently `darwin.apple_sdk_11_0.callPackage` uses a newer sdk on darwin and is a noop otherwise?

Correct, it's needed for x86_64-darwin support specifically, as aarch64-darwin already uses a newer SDK.

winterqt commented:
@ajs124 Can you change the CLANG_TIDY line to be unconditional, saying it doesn't work with clangStdenv (which Darwin uses by default)? I think that's better because it'll maintain support for both Darwin and clangStdenv on Linux users. (Also: the clang-tidy mechanism is a noop for non-Clang stdenv users right now anyways.)

ajs124 commented:
what about using `stdenv.cc.isClang`?

winterqt commented:
The CMake file doesn't even try to run clang-tidy with the gccStdenv (as it doesn't exist, so it silently disables it), but with the Clang stdenv, it'll see that clang-tidy exists, but the binary doesn't actually work because it's not wrapped.

winterqt commented on pkgs/applications/networking/p2p/transmission/default.nix:
```diff
@@ -59,11 +63,19 @@ in stdenv.mkDerivation {
+    ] ++ lib.optionals stdenv.isDarwin [
+      # Transmission sets this to 10.13 if not explicitly specified, see https://github.com/transmission/transmission/blob/0be7091eb12f4eb55f6690f313ef70a66795ee72/CMakeLists.txt#L7-L16.
+      "-DCMAKE_OSX_DEPLOYMENT_TARGET=${stdenv.hostPlatform.darwinMinVersion}"
+
+      # This doesn't work on Darwin, see #214945.
```
```suggestion
```

winterqt commented on pkgs/applications/networking/p2p/transmission/default.nix:
```diff
@@ -59,11 +63,19 @@ in stdenv.mkDerivation {
       "-DENABLE_DAEMON=${mkFlag enableDaemon}"
       "-DENABLE_CLI=${mkFlag enableCli}"
       "-DINSTALL_LIB=${mkFlag installLib}"
+      # breaks on darwin or with clangStdenv and is a noop otherwise
+      "-DRUN_CLANG_TIDY=OFF"
```
Keep styling?

```suggestion
      "-DRUN_CLANG_TIDY=OFF" # breaks on darwin or with clangStdenv and is a noop otherwise
```

winterqt reviewed (changes requested):
Guess we can just leave the alphabetical Foundation thing as is? *shrug*

ajs124 commented:
@vcunat @winterqt do you think this is good to go?

vcunat commented:
Do we mark `fragments` as broken?  It's trying to patch `transmission` which fails to apply now.  (otherwise it would build but no idea if it would work)

`torrential` breaks also, and it has maintainers which we haven't pinged (?): @xiorcale @davidak @bobby285271.

bobby285271 commented:
> torrential breaks also

Thanks for the ping. Not sure how to fix unfortunately, I filed https://github.com/davidmhewitt/torrential/issues/193.

ajs124 commented:
> Do we mark `fragments` as broken? It's trying to patch `transmission` which fails to apply now. (otherwise it would build but no idea if it would work)

It doesn't: https://github.com/NixOS/nixpkgs/pull/215316#issuecomment-1422906062

I tried updating it to the current release (2.1)  and dropping the patch, because AFAIR the patch only makes it so that transmission properly exits on a certain signal, which I think it does without patching with this release.

vcunat commented:
Well, I can imagine keeping `transmission_3` as non-default version used in these two packages, at least if they're still useful to someone.  (but I'd probably drop myself from `transmission_3.meta.maintainers`)

dmadisetti commented:
Transmission 4.0.1 has been released in this time (addressing minor bug fixes)

ajs124 commented:
I posted a version that keeps `transmission_3` around here: 17992eb1e451be84d6d34cce465d9e1c1e888293

It also updates to `4.0.1`.

If that's the preferred option, I'll push it to this branch.

jtojnar committed:
transmission: Clean up the expression

- Use one dependency per-line for cleaner diffs.
- Order derivation attributes according to convention.
- Extract apparmor rules expression into let binding.
- Use finalAttrs instead of let binding for consistent overriding.

jtojnar committed:
transmission_3: init at 3.00

Introduces pared down Transmission 3 package for apps that depend on it, until they are ported to Transmission 4.
Only keeping the library and daemon, as they are needed for Torrential and Fragments, respectively.
Not separating into different derivations them since the daemon only adds 1M to the closure.

jtojnar committed:
libutp: unstable-2017-01-02 → unstable-2023-03-05

https://github.com/transmission/libutp/compare/fda9f4b3db97ccb243fcbed2ce280eb4135d705b...9cb9f9c4f0073d78b08d6542cebaea6564ecadfe

Will be required for Transmission 4.0.0, keep the old version for other programs.

jtojnar commented:
@ajs124  Sounds good. Though since the transmission 3 is temporary, I would not bother making the expression generic and just clone it, it will be much cleaner that way. Also there is some opportunity for removing the third-party dependencies, see https://github.com/jtojnar/nixpkgs/commits/transmission

winterqt commented:
Can you please rebase?

ajs124 commented:
Which version are we going with now? https://github.com/jtojnar/nixpkgs/commit/440ab60f13e073a86dffbbf2f2a189fd840357b4 + https://github.com/jtojnar/nixpkgs/commit/f15c4187bfb6a486b6a56e8625657d587310c14c (but updated to 4.0.2)?

vcunat commented:
Copying the expression for `transmission_3` sounds OK to me, but I don't mind either way.

winterqt commented:
Given that, I'm inclined to wait at least until some activity happens in that issue.

ajs124 commented:
> can push it here as well if you want

iirc that doesn't work, since I opened this PR from a "shared" or "non-personal" repo. You can open a separate PR though, that supersedes this one, if you want to.

anna328p commented:
any updates on this?

vcunat commented:
I assumed we're waiting on that data-loss bug linked above.  (which has no updates)

anna328p commented:
somehow skipped over that message; good to know. i currently have an overlay updating transmission to 4.0.x in my configuration, will keep using that

chvp commented:
There doesn't seem to be a lot of movement on the bug linked above. Since the plan is to keep both transmission 3 and 4 around anyway, could we merge this but perhaps keep 3.0 as the default? I am personally looking to upgrade because of the performance improvements realized in 4.0. (I also don't know how common it is to skip downloading files, and the condition of zero-sized files having to be in the torrent as well makes it a rare-enough occurrence I would guess.)

Also, [4.0.3](https://github.com/transmission/transmission/releases/tag/4.0.3) got released a few weeks ago.

jtojnar commented on pkgs/top-level/all-packages.nix:
```diff
@@ -34884,15 +34884,16 @@ with pkgs;
   };
   transmission-gtk = transmission.override { enableGTK3 = true; };
   transmission-qt = transmission.override { enableQt = true; };
+  libtransmission = transmission_3;
```
This should probably go to `aliases.nix`. The only consumer, torrential already uses `transmission_3` directly.

jtojnar commented on pkgs/top-level/all-packages.nix:
```diff
@@ -34884,15 +34884,16 @@ with pkgs;
     enableDaemon = false;
     enableCli = false;
   };
   transmission-gtk = transmission.override { enableGTK3 = true; };
   transmission-qt = transmission.override { enableQt = true; };
```
These should be changed to use `transmission_3` too.

chvp commented on pkgs/top-level/all-packages.nix:
```diff
@@ -34882,14 +34884,16 @@ with pkgs;
-  libtransmission = transmission.override {
+  transmission_4 = callPackage ../applications/networking/p2p/transmission { };
+  transmission_3 = callPackage ../applications/networking/p2p/transmission/3.x.nix { };
+  libtransmission_4 = transmission.override {
```
This should be `transmission_4`, right?
```suggestion
  libtransmission_4 = transmission_4.override {
```

chvp committed:
transmission: fix eval

chvp commented:
I pushed a commit to my nixpkgs fixing eval: https://github.com/chvp/nixpkgs/commit/a49b75679d07185900963cb3168f8207118f42fc Feel free to cherry-pick/rebase into this branch.

ajs124 committed:
transmission_3: fix build

sveitser commented:
Have been using `transmission_4` from 68d19ed6aa236e1f045dfe80e494538f35df99a0 with the `transmission` nixos module for a few days and it has been working great.

chvp reviewed (approved):
Using the full patch as it's in this PR now, transmission 4 is running fine on my server as well (on x86_64).

SuperSandro2000 commented on pkgs/applications/networking/p2p/transmission/3.x.nix:
```diff
@@ -0,0 +1,79 @@
+  meta = {
+    description = "Old version of libtransmission library for apps that depend on it";
+    homepage = "http://www.transmissionbt.com/";
+    license = lib.licenses.gpl2Plus; # parts are under MIT
```
Then the license should be both

SuperSandro2000 commented on pkgs/applications/networking/p2p/transmission/3.x.nix:
```diff
@@ -0,0 +1,79 @@
+    description = "Old version of libtransmission library for apps that depend on it";
+    homepage = "http://www.transmissionbt.com/";
+    license = lib.licenses.gpl2Plus; # parts are under MIT
+    maintainers = with lib.maintainers; [ ];
```
Who wants to?

ajs124 commented:
feel free to take over anyone, I don't have time for this anymore

chvp commented:
I've taken over in #241146.

doronbehar commented:
For all those interested, I propose a rename making `transmission = transmission_4` in https://github.com/NixOS/nixpkgs/pull/258058 .
